types/basetypes/dynamic_value.go

```diff
@@ -0,0 +1,151 @@
+func NewDynamicValue(value attr.Value) DynamicValue {
+	return DynamicValue{
+		value: value,
+		state: attr.ValueStateKnown,
+	}
+}
+
+// NewDynamicNull creates a Dynamic with a null value. The concrete value type returned to Terraform
+// from this value will be tftypes.DynamicPseudoType.
+func NewDynamicNull() DynamicValue {
+	return DynamicValue{
+		state: attr.ValueStateNull,
+	}
+}
+
+// NewDynamicUnknown creates a Dynamic with an unknown value. The concrete value type returned to Terraform
+// from this value will be tftypes.DynamicPseudoType.
+func NewDynamicUnknown() DynamicValue {
+	return DynamicValue{
+		state: attr.ValueStateUnknown,
+	}
+}
+
+// DynamicValue represents a dynamic value. Static types are always
+// preferable over dynamic types in Terraform as practitioners will receive less
+// helpful configuration assistance from validation error diagnostics and editor
+// integrations.
+type DynamicValue struct {
+	// value contains the known value, if not null or unknown.
+	value attr.Value
+
+	// state represents whether the value is null, unknown, or known. The
+	// zero-value is null.
+	state attr.ValueState
+}
+
+// Type returns DynamicType.
+func (v DynamicValue) Type(ctx context.Context) attr.Type {
+	return DynamicType{}
+}
+
+// ToTerraformValue returns the equivalent tftypes.Value for the DynamicValue.
+func (v DynamicValue) ToTerraformValue(ctx context.Context) (tftypes.Value, error) {
+	switch v.state {
+	case attr.ValueStateKnown:
+		return v.value.ToTerraformValue(ctx)
+	case attr.ValueStateNull:
+		return tftypes.NewValue(tftypes.DynamicPseudoType, nil), nil
+	case attr.ValueStateUnknown:
+		return tftypes.NewValue(tftypes.DynamicPseudoType, tftypes.UnknownValue), nil
+	default:
+		panic(fmt.Sprintf("unhandled Dynamic state in ToTerraformValue: %s", v.state))
+	}
+}
+
+// Equal returns true if the given attr.Value is also a DynamicValue and contains an equal underlying value as defined by its Equal method.
+func (v DynamicValue) Equal(o attr.Value) bool {
+	other, ok := o.(DynamicValue)
+	if !ok {
+		return false
+	}
+
+	if v.state != other.state {
+		return false
+	}
+
+	if v.state != attr.ValueStateKnown {
+		return true
+	}
+
+	return v.value.Equal(other.value)
+}
+
+// IsNull returns true if the underlying value in the DynamicValue represents a null value.
+func (v DynamicValue) IsNull() bool {
+	return v.state == attr.ValueStateNull
+}
+
+// IsUnknown returns true if the underlying value in the DynamicValue represents an unknown value.
+func (v DynamicValue) IsUnknown() bool {
+	return v.state == attr.ValueStateUnknown
+}
+
+// String returns a human-readable representation of the DynamicValue. The string returned here is not protected by any compatibility guarantees,
+// and is intended for logging and error reporting.
+func (v DynamicValue) String() string {
+	if v.IsUnknown() {
+		return attr.UnknownValueString
+	}
+
+	if v.IsNull() {
+		return attr.NullValueString
+	}
+
+	return v.value.String()
```

Comment:
Should I double check this isn't `nil`? Return a default string?

Reply:
Same suggestion here to prevent panics -- it can be an awkward "<unimplemented>"/"<unset>"/"<invalid>" or similar string like we have for "<null>" and "<unknown>". I have a slight worry about saying "nil" directly because folks may interpret that to mean the same thing as null or somehow valid.